### PR TITLE
Add storage adapter factory (#51) (#55) (#61) (#64)

### PR DESCRIPTION
**Pull request by Johnmeurt**
This PR provides factories for the storage adapter. This allows evaluation of the adapter in runtime.
With this internal update you can now:
 - use env var in configuration file of the bundle
 - declare your own storage adapter
 - use DSN for storage declaration (bonus)

To declare a custom adapter you should add a custom facotry and add the tag `prometheus_metrics_bundle.adapter_factory`. I also register the factory interface for auto configuration.

The storage factory is designed to manage their own options. They should extract from array and cast if necessary their options  and also manage default values before injection. I could change this parameter to a class to simplify the usage options (I don't know if this is relevant).  

I change the configuration to move `type` and `redis` under a new parameter `storage` and I deprecated those parameters. 
I add a variable parameter `storage.options` to allows additionnal parameters. 

The container will have a new parameter for debug `prometheus_metrics_bundle.storage`. The container parameters `prometheus_metrics_bundle.type` and `prometheus_metrics_bundle.redis` are still set for debug purpose but are not use anymore.

I am not sure about the `storage`. It provides all known parameters for redis and apcu but I feel documentation could be enough. This means those parameters will move under `storage.options` as variable parameters.

I also add support of DSN. The value extract from the dsn will override the defined parameters.
Here is a example:

```yaml
artprima_prometheus_metrics:
    # namespace is used to prefix the prometheus metrics
    namespace: myapp

    # type and redis are still available.

    # metrics backend
    storage:
        # DSN of the storage. All parsed values will override explicitly set parameters. Ex: redis://127.0.0.1?timeout=0.1
        url: ~
        
        # Known values: in_memory, apcu, redis
        type: in_memory
        
        # Available parameters used by redis
        host: 127.0.0.1
        port: 6379
        timeout: 0.1
        read_timeout: 10
        persistent_connections: false
        password: ~
        database: ~ # Int value used by redis adapter
        prefix: ~   # String value used by redis and apcu

        # A variable parameter to define additionnal options as key / value.
        options:
            foo: bar
```

### CODE REVIEW

File: DependencyInjection/Compiler/ResolveAdapterDefinitionPass.php [hunk 1]
```diff
@@ -4,40 +4,32 @@
...
-use Prometheus\Storage\Redis;
+use Artprima\PrometheusMetricsBundle\StorageFactory\FactoryRegistry;
 use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\Compiler\PriorityTaggedServiceTrait;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
 
 /**
- * ResolveAdapterDefinitionPass is a compilation pass that sets the metrics backend storage adapter.
+ * ResolveAdapterDefinitionPass is a compilation pass that register factories of adapter.
```
Comment (denisvmedia) [inline DependencyInjection/Compiler/ResolveAdapterDefinitionPass.php:14]:
```suggestion
 * ResolveAdapterDefinitionPass is a compilation pass that registers adapter factories.
```

File: Tests/DependencyInjection/ConfigurationTest.php [hunk 2]
```diff
@@ -48,18 +49,19 @@ public function configDataProvider(): array
                 [
                     'namespace' => 'myapp',
                     'type' => 'redis',
-                    'ignored_routes' => ['prometheus_bundle_prometheus'],
                     'redis' => [
                         'host' => '127.0.0.1',
                         'port' => 6379,
                         'timeout' => 0.1,
-                        'read_timeout' => 10,
+                        'read_timeout' => '10',
```
Comment (denisvmedia) [inline Tests/DependencyInjection/ConfigurationTest.php:56]: What is the reason of this change?

File: StorageFactory/FactoryRegistry.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        // Only parsing dsn as url: legacy use case.
+        // This will manage non url value as scheme.
+        if (false === strpos($dsn, ':')) {
+            return ['scheme' => $dsn];
+        }
+
+        $options = parse_url($dsn);
+
+        if (false === $options) {
+            throw new InvalidArgumentException('Invalid DSN.');
```
Comment (denisvmedia) [inline StorageFactory/FactoryRegistry.php:69]: Probably we should include the given dsn in the error.

File: StorageFactory/StorageFactoryInterface.php [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Artprima\PrometheusMetricsBundle\StorageFactory;
+
+use Prometheus\Storage\Adapter;
+
+interface StorageFactoryInterface
```
Comment (denisvmedia) [inline StorageFactory/StorageFactoryInterface.php:9]: Do you mind giving it a short description on top of the interface?

File: Tests/DependencyInjection/ConfigurationTest.php (hunk 2 above)
Comment (Johnmeurt) [reply Tests/DependencyInjection/ConfigurationTest.php:56]: I add assertSame to ensure the type of values. The read_timeout was already cast in string to avoid an issue with Redis::setOption()

File: Tests/DependencyInjection/ConfigurationTest.php [hunk 5]
```diff
@@ -22,6 +22,7 @@ public function configDataProvider(): array
                 [
                     'namespace' => 'myapp',
                     'type' => 'in_memory',
+                    'storage' => ['type' => 'in_memory'],
```
Comment (denisvmedia) [inline Tests/DependencyInjection/ConfigurationTest.php:25]: I assume this is a test for the legacy use case. If so, why do we need it here?

File: Tests/DependencyInjection/ConfigurationTest.php [hunk 6]
```diff
@@ -48,18 +49,19 @@ public function configDataProvider(): array
...
                         'timeout' => 0.1,
-                        'read_timeout' => 10,
+                        'read_timeout' => '10',
                         'persistent_connections' => false,
                         'password' => null,
                     ],
                     'disable_default_metrics' => false,
                     'disable_default_promphp_metrics' => false,
                     'enable_console_metrics' => false,
+                    'storage' => ['type' => 'redis'],
```
Comment (denisvmedia) [inline Tests/DependencyInjection/ConfigurationTest.php:63]: The same comment here: why do we need to have a storage defined on a legacy use case? I'd assume that people who've just upgraded, shouldn't change their configuration.

File: Tests/DependencyInjection/ConfigurationTest.php (hunk 5 above)
Comment (Johnmeurt) [reply Tests/DependencyInjection/ConfigurationTest.php:25]: I let the configuration manage the legacy part. All values are inject into `storage` and the extension use `storage` it.
This behavior is internal and it will be easier to remove those deprecated parameters. I can remove this and let the extension manage the legacy but as I said it is internal.

File: Tests/DependencyInjection/ConfigurationTest.php (hunk 6 above)
Comment (Johnmeurt) [reply Tests/DependencyInjection/ConfigurationTest.php:63]: This change is on the expected part before compilation. We don't need to specify the storage. The configuration will inject values of `type` and `redis` to simplify usage in the extension that is why the storage appears in expected. The legacy config will still works fine.
